examples/pytorch/compgcn/model.py

```diff
@@ -0,0 +1,371 @@
+                    rel_graph.dstdata['comp_h'] = th.matmul(rel_graph.dstdata['comp_h'], self.W_I)
+                    if self.bias:
+                        rel_graph.dstdata['comp_h'] = rel_graph.dstdata['comp_h'] + self.b_I
+
+                # add composition results to each node's list
+                n_outputs[dtype].append(rel_graph.dstdata['comp_h'])
+
+            # self loop
+            self_rsts = {}
+            for ntype in g.ntypes:
+                self_rsts[ntype] = th.matmul(g.nodes[ntype].data['h'], self.W_S)
+                if self.bias:
+                    self_rsts[ntype] = self_rsts[ntype] + self.b_S
+
+
+        # aggregate all relations' composition results
+        n_out_feats = {}
+        for ntype, alist in n_outputs.items():
+            if len(alist) != 0:
+                stacked = th.stack(alist, dim=0)
+                n_out_feats[ntype] = th.sum(stacked, dim=0)
+
+                # add self loop results and dropout
+                n_out_feats[ntype] = self.dropout(n_out_feats[ntype]) * 2/3 + \
+                                     self.dropout(self_rsts.get(ntype)) * 1/3
+
+                # add batchnorm
+                if self.batchnorm:
+                    bn = nn.BatchNorm1d(self.hid_dim)
+                    n_out_feats[ntype] = bn(n_out_feats[ntype])
+
+                # use activate to compute non-linear
+                if self.activation is not None:
+                    n_out_feats[ntype] = self.activation(n_out_feats[ntype])
+
+        # compute edge embedding of the next layer
+        e_out_feats = {}
+        for etype, feat in e_in_feats.items():
+            e_out_feats[etype] = th.matmul(e_in_feats[etype], self.W_rel) + self.b_rel
+
+            # use activate to compute non-linear
+            if self.activation is not None:
+                e_out_feats[etype] = self.activation(e_out_feats[etype])
+
+        return n_out_feats, e_out_feats
+
+
+class CompGCN(nn.Module):
+    """
+    1. Will use dgl.nn.WeightBasis module to create basis vector of edge；
+    2.
+
+    """
+    def __init__(self,
+                 in_feat_dict,
+                 hid_dim,
+                 num_layers,
+                 out_feat,
+                 num_basis,
+                 num_rel,
+                 comp_fn='sub',
+                 dropout=0.0,
+                 activation=None,
+                 batchnorm=False
+                 ):
+        super(CompGCN, self).__init__()
+        self.in_feat_dict = in_feat_dict
+        self.rel_emb_dim = hid_dim
+        self.num_layer = num_layers
+        self.out_feat = out_feat
+        self.num_basis = num_basis
+        self.num_rel = num_rel
+
+        self.comp_fn = comp_fn
+        self.dropout = dropout
+        self.activation = activation
+        self.batchnorm = batchnorm
+
+        # Input layer, complete two tasks
+        # 1. define basis layer
+        self.basis = dglnn.WeightBasis([self.rel_emb_dim], self.num_basis, self.num_rel)
```

Comment:
what is this? is it the same as embedding layer?

Reply:
This is the basis vectors that are used to create z_r in formula 4 in the original paper.